# metatranscriptomics.py

PATH_KO_DB = '/home/pszczesny/soft/KEGG/ko.db'
PATH_KO_PCKL = '/home/pszczesny/soft/bipype/kogenes.pickle'

# refseq_bipype.py

PATH_GI_TAX = '/home/pszczesny/workingdata/refseq/db/tax_id/gi_taxid_nucl.dmp'
PATH_TAX_NAME = '/home/pszczesny/workingdata/refseq/db/tax_id/names.dmp'
PATH_RAPSEARCH = '/home/pszczesny/soft/RAPSearch2.12_64bits/bin/rapsearch'
PATH_REF_PROT_MASL28 = '/home/pszczesny/storage/workingdata/rapsearch/masl28'
PATH_REF_PROT_KO = '/home/pszczesny/soft/KEGG/ko.pep.rapsearch.db'
PATH_REF_PROT_ELSE = '/home/pszczesny/workingdata/refseq/protein/refseq_protein'
PATH_USEARCH = '/home/pszczesny/soft/usearch'
PATH_FQ2FA = '/usr/local/bioinformatics/fastx_toolkit/bin/fastq_to_fasta'
PATH_HUMANN = '/home/pszczesny/soft/humann-0.99'
PATH_HUMANN_DATA = '/home/pszczesny/soft/humann-0.99/input/hmp_metadata.dat'
PATH_BOWTIE2 = '/home/pszczesny/bin/bowtie2'
PATH_VELVETH = '/home/pszczesny/soft/Velvet/velveth'
PATH_VELVETG = '/home/pszczesny/soft/Velvet/velvetg'
PATH_METAVELVETG = '/home/pszczesny/soft/MetaVelvet/meta-velvetg'

# bipype

PREF_PATH_REF_FUNGI = '/home/pszczesny/workingdata/refseq/db/fungi/fungi_refseq'
PREF_PATH_REF_PLANT_1 = '/home/pszczesny/workingdata/refseq/db/plant/plant_refseq_1'
PREF_PATH_REF_PLANT_2 = '/home/pszczesny/workingdata/refseq/db/plant/plant_refseq_2'
PATH_X16S_DB = '/home/public/EMIRGE/SSU/SSU_candidate_db.udb'
PATH_ITS_DB = '/home/pszczesny/workingdata/ITS/unite.udb'
PATH_NCBI_TAXA_DB = '/home/pszczesny/soft/bipype/NCBI_tax_id_name.pickle'
PATH_RECONSTRUCT_DB = '/home/pszczesny/storage/workingdata/mcra/mcra10k.fasta'
PATH_16S_DATABASE = '/home/pszczesny/soft/bipype/SSU_candidate_db.fasta'
PATH_ITS_DATABASE = '/home/pszczesny/soft/bipype/UNITE_public_from_27.01.13.fasta'
